src/lib/blockdev.c

```diff
@@ -221,29 +237,107 @@ gboolean bd_init (BDPluginSpec **require_plugins, BDUtilsLogFunc log_func, GErro
+        if (require_plugins)
+            for (check_plugin=require_plugins; !missing && *check_plugin; check_plugin++)
+                missing = !bd_is_plugin_available((*check_plugin)->name);
+        else
+            /* all plugins requested */
+            for (plugin=BD_PLUGIN_LVM; plugin != BD_PLUGIN_UNDEF; plugin++)
+                missing = !bd_is_plugin_available(plugin);
+
+        if (!missing) {
+            g_mutex_unlock (&init_lock);
+            return TRUE;
+        }
     }
 
     if (log_func && !bd_utils_init_logging (log_func, error))
         /* the error is already populated */
         success = FALSE;
 
-    if (!load_plugins (require_plugins, FALSE)) {
+    if (!load_plugins (require_plugins, FALSE, &num_loaded)) {
         g_set_error (error, BD_INIT_ERROR, BD_INIT_ERROR_PLUGINS_FAILED,
                      "Failed to load plugins");
-        /* the library is unusable without the plugins so we can just return here */
         success = FALSE;
     }
 
-    initialized = success;
+    if (num_loaded == 0) {
+        if (require_plugins && (*require_plugins == NULL))
+            /* requested to load no plugins (NULL is the first item in the
+               array), none loaded -> OK */
+            initialized = TRUE;
+        else
+            initialized = FALSE;
+    } else
+        initialized = TRUE;
+
+    g_mutex_unlock (&init_lock);
+
+    return success;
+}
+
+/**
+ * bd_try_init:
+ * @request_plugins: (allow-none) (array zero-terminated=1): %NULL-terminated list
+ *                   of plugins that should be loaded (if no so_name is specified
+ *                   for the plugin, the default is used) or %NULL to load all
+ *                   plugins
+ * @log_func: (allow-none) (scope notified): logging function to use
+ * @loaded_plugin_names: (allow-none) (out) (transfer container): names of the successfully
+ *                                                                loaded plugins
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the library was successfully initialized with all the
+ *          required or default (see @require_plugins) plugins or not
+ *
+ * *UNLIKE IN CASE OF bd_init() AND bd_ensure_init(), FAILURE TO LOAD A PLUGIN
+ *  IS NOT CONSIDERED ERROR*
+ */
+gboolean bd_try_init(BDPluginSpec **request_plugins, BDUtilsLogFunc log_func,
+                     gchar ***loaded_plugin_names, GError **error) {
+    gboolean success = TRUE;
+    guint64 num_loaded = 0;
+
+    g_mutex_lock (&init_lock);
+    if (initialized) {
+        g_warning ("bd_init() called more than once! Use bd_reinit() to reinitialize "
+                   "or bd_is_initialized() to get the current state.");
+        g_mutex_unlock (&init_lock);
+        return FALSE;
+    }
+
+    if (log_func && !bd_utils_init_logging (log_func, error))
+        /* the error is already populated */
+        success = FALSE;
+
+    success = load_plugins (request_plugins, FALSE, &num_loaded);
```

Comment:
It would make more sense to me if ```load_plugins``` had an array of plugins as its third parameter, not a number, which is an information poor summary of the result of the loading effort. I think that ultimately it would be possible to be more direct in calling code with that interface.